src/settings/settings_manager.cpp

```diff
@@ -0,0 +1,227 @@
+    } else {
+      ActionState action_state =
+          InvokeCallback(param, static_cast<void *>(&old_value), static_cast<void *>(&value), action_context);
+      if (action_state == ActionState::FAILURE) {
+        TERRIER_ASSERT(SetValue(param, ValueFactory::GetInteger(old_value)),
+                       "Resetting parameter value should not fail");
+      }
+    }
+  }
+  setter_callback(action_context);
+}
+
+void SettingsManager::SetDouble(Param param, double value, std::shared_ptr<ActionContext> action_context,
+                                setter_callback_fn setter_callback) {
+  common::SharedLatch::ScopedExclusiveLatch guard(&latch_);
+  const auto &param_info = db_->param_map_.find(param)->second;
+  double min_value = param_info.min_value;
+  double max_value = param_info.max_value;
+  if (!(value >= min_value && value <= max_value)) {
+    action_context->SetState(ActionState::FAILURE);
+  } else {
+    double old_value = ValuePeeker::PeekDecimal(GetValue(param));
+    if (!SetValue(param, ValueFactory::GetDecimal(value))) {
+      action_context->SetState(ActionState::FAILURE);
+    } else {
+      ActionState action_state =
+          InvokeCallback(param, static_cast<void *>(&old_value), static_cast<void *>(&value), action_context);
+      if (action_state == ActionState::FAILURE) {
+        TERRIER_ASSERT(SetValue(param, ValueFactory::GetDecimal(old_value)),
+                       "Resetting parameter value should not fail");
+      }
+    }
+  }
+  setter_callback(action_context);
+}
+
+void SettingsManager::SetBool(Param param, bool value, std::shared_ptr<ActionContext> action_context,
+                              setter_callback_fn setter_callback) {
+  common::SharedLatch::ScopedExclusiveLatch guard(&latch_);
+  bool old_value = ValuePeeker::PeekBoolean(GetValue(param));
+  if (!SetValue(param, ValueFactory::GetBoolean(value))) {
+    action_context->SetState(ActionState::FAILURE);
+  } else {
+    ActionState action_state =
+        InvokeCallback(param, static_cast<void *>(&old_value), static_cast<void *>(&value), action_context);
+    if (action_state == ActionState::FAILURE) {
+      TERRIER_ASSERT(SetValue(param, ValueFactory::GetBoolean(old_value)), "Resetting parameter value should not fail");
+    }
+  }
+  setter_callback(action_context);
+}
+
+void SettingsManager::SetString(Param param, const std::string_view &value,
+                                std::shared_ptr<ActionContext> action_context, setter_callback_fn setter_callback) {
+  common::SharedLatch::ScopedExclusiveLatch guard(&latch_);
+  std::string_view old_value = ValuePeeker::PeekVarChar(GetValue(param));
+  if (!SetValue(param, ValueFactory::GetVarChar(value))) {
+    action_context->SetState(ActionState::FAILURE);
+  } else {
+    std::string_view new_value(value);
+    ActionState action_state =
+        InvokeCallback(param, static_cast<void *>(&old_value), static_cast<void *>(&new_value), action_context);
```

Comment:
If you are passing a pointer to function with argument type `void *`, IIRC there is no need to use a cast.

Reply:
That is news to me. Thank you!